server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java

```diff
@@ -73,554 +49,125 @@
-        }
-        if (fieldType().strategyName().equals(SpatialStrategy.TERM.getStrategyName())) {
-            // For TERMs strategy the defaults for points only change to true
-            if (includeDefaults || fieldType().pointsOnly() != true) {
-                builder.field(Names.STRATEGY_POINTS_ONLY, fieldType().pointsOnly());
+                    fieldType().typeName());
+            }
+            context.addIgnoredField(fieldType().name());
+        }
+    }
+
+    private void indexShape(ParseContext context, Object luceneShape) {
+        if (luceneShape instanceof GeoPoint) {
+            GeoPoint pt = (GeoPoint)luceneShape;
+            indexFields(context, LatLonShape.createIndexableFields(name(), pt.lat(), pt.lon()));
+        } else if (luceneShape instanceof Line) {
+            indexFields(context, LatLonShape.createIndexableFields(name(), (Line)luceneShape));
+        } else if (luceneShape instanceof Polygon) {
+            indexFields(context, LatLonShape.createIndexableFields(name(), (Polygon) luceneShape));
+        } else if (luceneShape instanceof double[][]) {
+            double[][] pts = (double[][])luceneShape;
+            for (int i = 0; i < pts.length; ++i) {
+                indexFields(context, LatLonShape.createIndexableFields(name(), pts[i][1], pts[i][0]));
+            }
+        } else if (luceneShape instanceof Line[]) {
+            Line[] lines = (Line[]) luceneShape;
+            for (int i = 0; i < lines.length; ++i) {
+                indexFields(context, LatLonShape.createIndexableFields(name(), lines[i]));
+            }
+        } else if (luceneShape instanceof Polygon[]) {
+            Polygon[] polys = (Polygon[]) luceneShape;
+            for (int i = 0; i < polys.length; ++i) {
+                indexFields(context, LatLonShape.createIndexableFields(name(), polys[i]));
+            }
+        } else if (luceneShape instanceof Rectangle) {
+            // index rectangle as a polygon
+            Rectangle r = (Rectangle) luceneShape;
+            Polygon p = new Polygon(new double[]{r.minLat, r.minLat, r.maxLat, r.maxLat, r.minLat},
+                new double[]{r.minLon, r.maxLon, r.maxLon, r.minLon, r.minLon});
+            indexFields(context, LatLonShape.createIndexableFields(name(), p));
+        } else if (luceneShape instanceof Object[]) {
+            // recurse to index geometry collection
+            for (Object o : (Object[])luceneShape) {
+                indexShape(context, o);
             }
         } else {
-            if (includeDefaults || fieldType().pointsOnly() != GeoShapeFieldMapper.Defaults.POINTS_ONLY) {
-                builder.field(Names.STRATEGY_POINTS_ONLY, fieldType().pointsOnly());
-            }
-        }
-        if (includeDefaults || coerce.explicit()) {
-            builder.field(Names.COERCE, coerce.value());
+            throw new IllegalArgumentException("invalid shape type found [" + luceneShape.getClass().getName() + "] while indexing shape");
```

Comment:
getClass().getName() will not be meaningful eg. for an anonymous class, let's just rely on the toString representation of the class?

Reply:
:+1: